runtime/src/iree/hal/utils/libmpi.c

```diff
@@ -67,21 +69,39 @@ iree_status_t iree_hal_mpi_library_load(
 // convert an MPI_Status result to an iree_status_t
 //
 // We convert MPI_SUCCESS to IREE_STATUS_SUCCESS, and MPI errors to
-// IREE_STATUS_INTERNAL and print the MPI error code.
-//
-// TODO(cascaval): use MPI_Error to print the actual error or consider
-// mapping to other error codes.
+// IREE_STATUS_INTERNAL and retrieve the error string from the MPI library.
 iree_status_t iree_hal_mpi_result_to_status(
     iree_hal_mpi_dynamic_symbols_t* syms, int result, const char* file,
     uint32_t line) {
-  iree_status_code_t code;
-  switch (result) {
-    case 0:  // MPI_SUCCESS
-      return iree_ok_status();
-    default:
-      code = IREE_STATUS_INTERNAL;
-      break;
+  // the MPI spec mandates MPI_SUCCESS == 0, all the other errors are
+  // implementation dependent. However, since we already depend on OpenMPI
+  // for implementation dependent handles, we can hardcode the error classes
+  // as well.
+  const int MPI_SUCCESS = 0;
+  const int MPI_ERR_UNKNOWN = 14;
+#define MPI_MAX_ERROR_STRING 256
+
+  if (IREE_LIKELY(result == MPI_SUCCESS)) {
+    return iree_ok_status();
   }
-  return iree_make_status_with_location(file, line, code, "MPI error %d",
-                                        result);
+
+  if (!syms) {
+    return iree_make_status_with_location(file, line, IREE_STATUS_INTERNAL,
+                                          "MPI library symbols not loaded");
+  }
+
+  char error_string[MPI_MAX_ERROR_STRING];
+  int error_len = 0;
+  if (syms->MPI_Error_string(result, error_string, &error_len) != MPI_SUCCESS) {
+    strncpy(error_string, "UNKNOWN", 8);
```

Comment:
```suggestion
    strncpy(error_string, "UNKNOWN", strlen("UNKNOWN"));
```
(or leave as empty string if there's nothing meaningful)

Reply:
left empty.